akrypt/akrypt.1
.\" Automatically generated by Pandoc 2.2.1
.\"
.TH "akrypt" "1" "18 июля 2019 г." "Русский мануал для программы akrypt" ""
.hy
.SH НАЗВАНИЕ
.PP
akrypt \- консольная утилита для выполнения криптографических
преобразований.
.SH СИНТАКСИС
.PP
Синтаксис выполнения программы
.PP
akrypt команда [\f[I]опции\f[]] [\f[I]файл\f[] или \f[I]каталог\f[]]
\&...
.SH ОПИСАНИЕ
.PP
Утилита предназначена для работы с ключами и криптографическими
преобразованиями, реализуемыми библиотекой \f[I]libakrypt\f[], включая
контроль целостности файлов, шифрование данных, а также функции для
работы с ключевой информацией.
.PP
Каждая команда определяет отдельный класс криптографических
преобразований и функциональных возможностей программы.
В настоящее время доступны следующие команды:
.PP
\f[I]icode\f[] Вычисление контрольной суммы или имитовставки для
заданного файла
.PP
\f[I]show\f[] Вывод служебной и справочной информации
.PP
Получить подробную информацию о доступных опциях и параметрах любой
команды можно с помощью вызова
.IP
.nf
\f[C]
akrypt\ команда\ \-\-help
\f[]
.fi
.SH КОМАНДЫ ДЛЯ КОНТРОЛЯ ЦЕЛОСТНОCТИ
.SS i [\f[I]опции\f[]] \f[I]файл\f[] или \f[I]каталог\f[]
.PP
Короткая форма команды \f[I]icode\f[].
.SS icode [\f[I]опции\f[]] \f[I]файл\f[] или \f[I]каталог\f[]
.PP
Команда icode позволяет вычислять контрольные суммы или имитовставки для
одного или нескольких файлов.
В случае указания в командной строке имени каталога, контрольные суммы
или имитовставки будут подсчитаны для всех доступных для чтения файлов в
указанном каталоге.
.PP
При вычислении имитовставки должен быть использован секретный ключ.
Данный ключ должен быть создан заранее, либо выработан из заданного
пользователем пароля.
.PP
В настоящий момент доступны следующие опции.
.TP
.B \-a, \-\-algorithm <\f[I]ni\f[]>
Опция позволяет указать задаваемый параметром \f[I]ni\f[] (name or
identifier) алгоритм, который будет использован для вычисления
контрольной суммы или имитовставки.
В случае, когда опция не определена, для вычисления контрольной суммы
используется функция хеширования \[lq]Стрибог256\[rq],
регламентированная национальным стандартом ГОСТ Р 34.11\-2012.
Получить перечень всех доступных алгоритмов можно с помощью команды
\f[I]show\f[], см.
ниже.
.RS
.RE
.TP
.B \\
При вычислении имитовставок обязательно использование секретного ключа.
Поскольку некоторые алгоритмы накладывают ограничения на объем
информации, обрабатываеой на одном ключе, то для таких алгоритмов
вычисление имитоставок от длинных (больших) файлов может быть
недоступно.
В этом случае вместо имитовставки выводится значение \f[I]skipped\f[].
Если Вам необходимо вычислять имитовставки файлов с произвольной, сколь
угодно большой длиной, то используйте следующие алгоритмы:
\f[I]hmac\-streebog256\f[] и \f[I]hmac\-streebog512\f[].
.RS
.RE
.TP
.B \-c, \-\-check <\f[I]file\f[]>
Опция позволяет проверить контрольные суммы или имитовставки для одного
или нескольких файлов.
Данные суммы должны быть вычислены заранее и сохраненны в файле
\f[I]file\f[].
Если при проверке используется алгоритм, отличный от установленного
по\-умолчанию, необходимо указать его имя или идентификатор с помощью
опции \f[I]\-a\f[].
.RS
.RE
.TP
.B \-o, \-\-output <\f[I]file\f[]>
Опция определяет имя файла, в который записываются вычисленные значения
контрольных сумм или имитовставок.
Данный файл может быть указан в качестве параметра опции \f[I]\-c\f[].
.RS
.RE
.TP
.B \-p
Опция указывает, что для выработки секретного ключа, используемого при
вычислении или проверке имитовставки от одного или нескольких файлов,
должен использоваться задаваемый пользователем пароль.
Если опция \f[I]password\f[] не опредена, то программа предлагает
пользователю ввести пароль с консоли.
.RS
.RE
.TP
.B \-\-password <\f[I]pass\f[]>
Опция в явном виде указывает, что для выработки секретного ключа должен
использоваться пароль, определяемый последовательностью символов
\f[I]pass\f[].
Использование данной опции не является безопасным.
.RS
.RE
.TP
.B \-r, \-\-recursive
Опция указывает, что при вычислении контрольных сумм или имитовставок,
должна выполняться рекурсивная процедура обхода текущего и всех
вложенных каталогов.
.RS
.RE
.TP
.B \-\-reverse\-order
Опция указывает, что все выводимые и вводимые последовательности
октетов, должны обрабатываться в обратном порядке.
Использование данной опции целесообразно для вывода результатов в big
endian порядке следования октетов.
.RS
.RE
.TP
.B \-\-tag
Опция указывает, что значения контрольных сумм или имитовставок должны
выводиться в формате, принятом в операционных системах семейства BSD.
В таком формате указывается не только имя файла и его контрольная сумма,
но и алгоритм, с помощью которого данная сумма была подсчитана.
.RS
.RE
.TP
.B \-t, \-\-template
Опция указывает шаблон поиска файлов.
Для задания шаблона используются правила, аналогичные правилам,
определенным для функции \f[I]fnmatch\f[].
.RS
.RE
.PP
Следующие опции имеет смысл применять только при проверке контрольных
сумм или имитовставок.
.TP
.B \-\-dont\-show\-stat
Опция запрещает вывод в консоль статистической информации об общем
количестве проверенных файлов, количестве успешных или неуспешных
проверок и т.п.
.RS
.RE
.TP
.B \-\-ignore\-errors
Опция запрещает останавливать процесс проверки контрольных сумм или
имитовставок в случае возникновения ошибок доступа или чтения файлов,
отсутствия файлов, содержащихся в списке и т.п.
.RS
.RE
.TP
.B \-\-quiet
Опция запрещает вывод символов \[lq]Ok\[rq] после успешной проверки
контрольной суммы.
.RS
.RE
.TP
.B \-\-status
Опция запрещает вывод какой\-либо информации при проверке контрольных
сумм.
Результат проверки возвращается в коде возврата программы (любое
отличное от нуля значение сигнализирует об ошибке).
.RS
.RE
.SH ВСПОМОГАТЕЛЬНЫЕ И ИНФОРМАЦИОННЫЕ КОМАНДЫ
.SS show [\f[I]опции\f[]]
.PP
Команда позволяет получить информацию о доступных криптографических
преобразованиях, значениях технических и криптографических
характеристик, параметрах библиотеки по\-умолчанию.
В настоящий момент доступны следующие опции.
.TP
.B \-\-engines
Опция позволяет вывести список всех типов криптографических механизмов,
которые могут быть использованы в приложениях библиотеки libakrypt.
К таким типам, например, относятся блочные шифры или алгоритмы выработки
имитовставки.
.RS
.RE
.TP
.B \-\-modes
Опция позволяет вывести список всех режимов криптографических
преобразований, которыми могут быть охарактеризованы криптографические
преобразования.
При этом одному криптографическому механизму может соответствовать
несколько режимов.
Например, собственно функция хэширования имеет режим algorithm, а ее
параметры \- kbox params.
.RS
.RE
.TP
.B \-\-oids
Опция позволяет вывести список всех доступных криптографических
механизмов.
Каждый механизм идентифицируется свои именем, которое используется для
его указания пользователем и/или последовательностью чисел, разделенных
точками.
Последняя последовательность называется Object IDentifier (OID) и
используется при автоматизированной обработке данных.
Например, среди прочих, будет выведена следующая информация об алгоритме
блочного шифрования Магма
.RS
.PP
\f[I]magma\f[] block cipher algorithm 1.2.643.7.1.1.5.1
.PP
в которой \f[I]magma\f[] является именем криптографического механизма, а
1.2.643.7.1.1.5.1 \- его идентификатором.
.RE
.TP
.B \-\-oid <\f[I]eni\f[]>
Опция принимает в качестве параметра произвольную строку символов
\f[I]eni\f[] (engine, name or identifier) и выводит все доступные
криптографические механизмы, в именах или идентификаторах которых
содержится указанная строка.
Например, вызов
.RS
.PP
akrypt show \[en]oid magma
.PP
может привести к следующему переченю криптографических механизмов.
.PP
\f[I]omac\-magma\f[] omac function algorithm 1.2.643.2.52.1.4.1
.PP
\f[I]mgm\-magma\f[] mgm function algorithm 1.2.643.2.52.1.4.3
.PP
\f[I]magma\f[] block cipher algorithm 1.2.643.7.1.1.5.1
.RE
.TP
.B \-\-options
Опция выводит перечень всех криптографических параметров библиотеки.
К указанным парметрам относятся, например, число блоков информации,
которые могут быть зашифрованы на одном ключе, или число итераций
алгоритма PBKDF2, используемого для развертки ключа из пароля.
.RS
.RE
.TP
.B \-\-without\-caption
Опция запрещает печать заголовка, расшифровывающего названия выводимых
параметров и их значений.
.RS
.RE
.SH ОПЦИИ
.SH ДОПОЛНИТЕЛЬНАЯ ИНФОРМАЦИЯ
.SH ПРИМЕРЫ ШИФРОВАНИЯ ИНФОРМАЦИИ
.SH ПРИМЕРЫ КОНТРОЛЯ ЦЕЛОСТНОСТИ
.SS akrypt i file.txt \-o result.streebog
.PP
Данный вызов вычисляет контрольную сумму файла file.txt с помощью
установленного по умолчанию алгоритма \[lq]Стребог256\[rq] и помещает
результат вычислений в файл result.streebog.
.SS akrypt i \-rt "*.t??" \-a omac\-kuznechik \-p .
.PP
Данный вызов позволяет вычислить код целостности всех файлов,
удовлетворяющих шаблону "*.t??" (файлы, имеющие расширение из трех
символов, начинающееся с символа t) в текущем каталоге (на это указывает
символ \[lq].\[rq]), а также во всех доступных вложенных каталогах.
Для вычисления кода целостности используется алгоритм выработки
имитовставки ГОСТ Р 34.13\-2015 в основе которого лежит блочный шифр
\[lq]Кузнечик\[rq].
Для имитозащиты файлов используется ключ, вырабатываемый из пароля,
который должен быть введен пользователем.
.SS akrypt i \-c result.streebog
.PP
Данный вызов позволяет проверить контрольные суммы, указанные в файле
result.streebog.
При проверке используется установленный по\-умолчанию алгоритм
\[lq]Стрибог256\[rq].
.SS akrypt i \-c result.txt \-a hmac\-streebog256 \[en]password aQ13jzUl
.PP
Данный вызов позволяет проверить значения имитовставок, указанных в
файле result.txt При проверке используется алгоритм hmac\-streebog256,
регламентированный Р 50.1.113\-2016.
Ключ имитозащиты вырабатывается из пароля, указанного пользователем в
командной строке в явном виде.
.SH СРАВНЕНИЕ С УТИЛИТАМИ С АНАЛОГИЧНОЙ ФУНКЦИОНАЛЬНОСТЬЮ
.SS Контроль целостности файлов
.PP
Известны две общедоступные утилиты, позволяющие вычислять контрольные
суммы с помощью отечественных алгоритмов хеширования.
Этими утилитами являются
.IP \[bu] 2
gost12sum ( https://github.com/gost\-engine/engine ),
.IP \[bu] 2
rhash ( https://github.com/rhash/RHash ).
.PP
Следующая последовательность команд позволяет продемонстрировать
эквивалентность работы всех трех утилит для алгоритма хеширования
Стрибог\-256.
.IP
.nf
\f[C]
echo\ \-n\ hello\ >\ test.file
gost12sum\ test.file
3fb0700a41ce6e41413ba764f98bf2135ba6ded516bea2fae8429cc5bdd46d6d\ test.file

rhash\ \-G\ test.file
3fb0700a41ce6e41413ba764f98bf2135ba6ded516bea2fae8429cc5bdd46d6d\ \ test.file

akrypt\ i\ test.file
3FB0700A41CE6E41413BA764F98BF2135BA6DED516BEA2FAE8429CC5BDD46D6D\ test.file
\f[]
.fi
.PP
Аналогично, для алгоритма хеширования Стрибог512, запуск следующих
команд также позволит получить одинаковое значение хэш\-кода размером
512 бит.
.IP
.nf
\f[C]
gost12sum\ \-l\ test.file
rhash\ \-\-gost12\-512\ test.file
akrypt\ i\ \-a\ streebog512\ test.file
\f[]
.fi
.SH СТАНДАРТЫ
.PP
Национальные стандарты Российской Федерации
.IP \[bu] 2
ГОСТ Р 34.10\-2012,
.IP \[bu] 2
ГОСТ Р 34.11\-2012,
.IP \[bu] 2
ГОСТ Р 34.12\-2015,
.IP \[bu] 2
ГОСТ Р 34.13\-2015.
.PP
Рекомендации по стандартизации Росстандарта России
.IP \[bu] 2
Р 50.1.111\-2016,
.IP \[bu] 2
Р 50.1.113\-2016
.SH ИНФОРМАЦИЯ О ПРОЕКТЕ
.PP
Сайт проекта http://libakrypt.org
.PP
Исходные коды проекта: https://github.com/axelkenzo/libakrypt\-0.x
.SH АВТОРЫ
.PP
Axel Kenzo & The Company Of Belles Lettres (с) 2014 \- 2019


Changelog.md
# Перечень изменений


## Изменения в версии 0.7.15

 - Добавлены функции умножения элементов в полях GF(2^256) и GF(2^512)
 - Оптимизирована реализация функций хеширования для big-endian архитектуры
 - Унифицирован вызов стандартных функций для работы с сокетами
 - В реализацию протокола sp fiot добавлена поддержка расширений requestIdentifier
 - Добавлена тестовая реализация ASN.1 парсера


## Изменения в версии 0.7.14

 - Изменения в функциях внешнего интерфейса:
   * добавлены функции для создания и применения объектов класса mac
   * удалены функции внешнего интерфейса для класса hash
   * реализованы функции для доступа к параметрам OID алгоритма по индексу и по дескриптору

 - Добавлены функции внутреннего инфтерфейса ak_mac_context_is_iv_settable() и ak_mac_context_get_iv_size()
 - Изменено внутреннее устройстро структуры mac,
   реализующей общий подход к реализации алгоритмов итерационного сжатия
 - Изменено внутреннее устройство структуры key_resource_counter, хранящей ресурс секретного ключа.
 - Факт установки ресурса ключа теперь привязан к моменту присвоения ключевого значения.
 - Добавлена краткая документация к утилите akrypt 


## Изменения в версии 0.7.13

 - Добавлена реализация режима работы блочных шифров:
   * ACPKM для шифрования данных ( функция ak_bckey_context_ctr_acpkm() )
 - Добавлен ряд служебных функций для работы с контекстами секретных ключей:
   * ak_skey_context_set_resource()
   * ak_skey_context_create_and_set_skey()
   * ak_bckey_context_create_and_set_bckey()
 - Название функции гаммирования для блочных шифров приведено к виду, соответствующему тексту стандарта,
   а именно ak_bckey_context_ctr()


## Изменения в версии 0.7.12

 - В состав библиотеки включена первая реализация протокола защищенного взаимодействия sp fiot.
 - Исправлена ошибка в реализации Кузнечика, из-за которой библиотека
   падала на некоторых комбинациях 32-х битной архитектуры, как правило Celeron, и компиляторов gcc и clang.
 - Исправлены баги в функции unmask() для алгоритма Магма (приведение указателя + разворот байт).
 - Исправлен баг при передаче длинных синхропосылок в ak_bckey_context_xcrypt().
 - Функция ak_buffer_alloc() теперь сохраняет предыдущие данные при увеличении длины буффера.
 - Добавлен ряд функций для работы с классом mac.
 - Добавлены функции сериализации/десериализации вычетов, используемые при хранении точек эллиптических кривых.
 - Добавлен генератор случайных чисел, предназначенный для считывания данных из сокетов домена unix.
 - Внесены небольшие изменения в документацию к библиотеке.
 - Уточнены процедуры инсталяции и деинсталяции библиотеки.
 - Возвращена консольная утилита akrypt (сборка утилиты может быть включена флагом cmake).


## Изменения в версии 0.7.11

 - Отлажена реализация блочных шифров "Кузнечик" и "Магма" для big endian архитектуры
 - Отлажена реализация режимов ecb, ctr, omac и mgm для big endian архитектуры
 - Исправлен баг шифрования коротких векторов в режимах xcrypt
 - Разрешен экспорт функции ak_log_set_level() (только для повышения уровня аудита)
 - Изменен флаг сборки библиотеки с --std=c11 на более ранний --std=c99
   (оказался нужен для поддержки ранних версий gcc)


## Изменения в версии 0.7.10

 - Начиная с этой версии не поддерживается алгоритм хеширования ГОСТ Р 34.11-94
 - Все остальные алгоритмы хеширования, а также основанные на них алгоритмы HMAC и PBKDF2,
   корректно выполняются на big endian архитектуре (протестированы arm32v7eb, mips32 и mips64)
 - Корректно выполняются на big endian архитектуре алгоритмы выработки и проверки электронной подписи.
 - Разделена сборка документации для различных форматов. Теперь доступны команды make html и make pdf.


## Изменения в версии 0.7.9

 - Сформировано ядро библиотеки - минимальный набор функций и тестовых примеров, позволяющий проводить
   тестирование процесса сборки без учета ошибок в реализации криптографических механизмов
 - Эллиптические кривые корректно тестируются под big endian архитектурой
 - Функции преобразования строк в большие целые числа  и обратно теперь учитывают архитектуру,
   на которой исполняется библиотека


## Изменения в версии 0.7.8

 - Исправлен ряд ошибок при сборке под x86_32 (компиляция asm вставок, тестирование архитектуры, поддержка Windows32 и т.д.)
 - Начата поддрежка big endian архитектуры (пока для операций в GF(2^n))


## Изменения в версии 0.7.7

 - Уточнена реализация алгоритма MGM,
   сделан вызов алгоритма выработки имитовставки MGM через интерфейс класса mac
 - Более аккуратно распределено включение стандартных заголовочных файлов

## Изменения в версии 0.7.6

 - Теперь сборка тестовых примеров производится только при установленном флаге
   -DLIBAKRYPT_INTERNAL_TESTS=ON при запуске cmake
 - Процедура тестирования криптографических механизмов вынесена в отдельную функцию
   ak_libakrypt_dynamic_control_test()
 - Немного оптимизирован механизм работы менеджера контекстов
 - Сделана корректная обработка рускоязычных имен файлов для консольного прилоджения akrypt (через gettext)
 - Исправлена ошибка выработки контрольной суммы для отсутствующих файлов (struct mac)


## Изменения в версии 0.7.5

 - Сделана реализация алгоритма выработки имитовставки ГОСТ Р 34.13-2015 через
   функции классы omac и mac
 - Исправлен ряд ошибок в реализации функций присвоения значений секретным ключам (класс skey)
 - Восстановлена реализация алгоритмов выработки и проверки электронной подписи.
 - Сделана возможность создания секретного ключа электронной подписи по идентификаторам (oid)
   алгоритма и эллиптической кривой.
 - Добавлен ряд тестов в механизм тестирования электронной подписи.
 - Переименование типов: ak_oid_engine -> oid_engines и ak_oid_mode -> oid_modes


## Изменения в версии 0.7.4

 - Изменена структура класса object, содержащая указатели на функции создания и
   удаления классов.
 - Исправлена реализация функции pbkdf2 и механизма генерации ключей из пароля (теперь

akrypt/akrypt.c
/* ----------------------------------------------------------------------------------------------- */
/*  Copyright (c) 2014 - 2019 by Axel Kenzo, axelkenzo@mail.ru                                     */
/*                                                                                                 */
/*  Файл akrypt.c                                                                                  */
/*  - содержит реализацию консольной утилиты, иллюстрирующей возможности библиотеки libakrypt      */
/* ----------------------------------------------------------------------------------------------- */
 #include <time.h>
 #include <stdio.h>
 #include <string.h>
 #include <stdlib.h>
 #include <akrypt.h>
#ifdef _MSC_VER
 #include <strsafe.h>
#endif

/* ----------------------------------------------------------------------------------------------- */
 char audit_filename[1024];
/* функция, реализующая аудит */
 ak_function_log *audit =
#ifdef _WIN32
 akrypt_audit_function;
#else
 ak_function_log_syslog;
#endif

/* ----------------------------------------------------------------------------------------------- */
 int main( int argc, TCHAR *argv[] )
{
 /* определение переменных, используемых ддя указания времени старта программы */
 #ifdef _WIN32
  time_t ptime;
  TCHAR homepath[FILENAME_MAX]
   #ifdef MSC_VER
     , buffer[64]
   #endif
   ;
 #endif
 /* попытка русификации программы для unix-like операционных систем */
 #ifdef LIBAKRYPT_HAVE_LIBINTL_H
 /* обрабатываем настройки локали
    при инсталляции файл akrypt.mo должен помещаться в /usr/share/locale/ru/LC_MESSAGES */
  #ifdef LIBAKRYPT_HAVE_LOCALE_H
   setlocale( LC_ALL, "" );
  #endif
  bindtextdomain( "akrypt", "/usr/share/locale/" );
  textdomain( "akrypt" );
 #endif

 /* проверяем, что пользователем должна быть задана команда */
  if( argc < 2 ) return akrypt_litehelp();

 /* проверяем флаги вывода справочной информации */
  if( akrypt_check_command( "-h", argv[1] )) return akrypt_help();
  if( akrypt_check_command( "--help", argv[1] )) return akrypt_help();
  if( akrypt_check_command( "/?", argv[1] )) return akrypt_help();

 #ifdef _WIN32
  if( ak_libakrypt_get_home_path( homepath, FILENAME_MAX ) == ak_error_ok ) {
    ak_snprintf( audit_filename, FILENAME_MAX, "%s\\.config\\libakrypt\\akrypt.log", homepath );
    remove( audit_filename );
    ak_log_set_function( audit = akrypt_audit_function );

   #ifdef MSC_VER
    _time64( &ptime );
    _tctime64_s( buffer, sizeof( buffer ), &ptime );
    ak_snprintf( homepath, FILENAME_MAX, "%s started at %s", argv[0], buffer );
   #else
     ptime = time( NULL );
     ak_snprintf( homepath, FILENAME_MAX, "%s started at %s", argv[0], ctime( &ptime ));
   #endif
    ak_log_set_message( homepath );
  }

  SetConsoleCP( 1251 );
  SetConsoleOutputCP( 1251 );
 #endif

 /* выполняем команду пользователя */
  if( akrypt_check_command( "show", argv[1] )) return akrypt_show( argc, argv );
  if( akrypt_check_command( "i", argv[1] )) return akrypt_icode( argc, argv );
  if( akrypt_check_command( "icode", argv[1] )) return akrypt_icode( argc, argv );

  if( akrypt_check_command( "mgm", argv[1] )) return akrypt_mgm( argc, argv );


 /* ничего не подошло, выводим сообщение об ошибке */
  ak_log_set_function( ak_function_log_stderr );
  ak_error_message_fmt( ak_error_undefined_function,
                                                 __func__, _("undefined command \"%s\""), argv[1] );
 return EXIT_FAILURE;
}

/* ----------------------------------------------------------------------------------------------- */
/*                         реализация функций обработки команд пользователя                        */
/* ----------------------------------------------------------------------------------------------- */
 bool_t akrypt_check_command( const char *comm, TCHAR *argv )
{
 size_t len = strlen( comm );

  if( strlen( argv ) != len ) return ak_false;
  if( strncmp( comm, argv, len )) return ak_false;
 return ak_true;
}

/* ----------------------------------------------------------------------------------------------- */
/* вывод сообщений в заданный пользователем файл, а также в стандартный демон вывода сообщений */
 int akrypt_audit_function( const char *message )
{
  FILE *fp = fopen( audit_filename, "a+" );
   /* функция выводит сообщения в заданный файл */
    if( !fp ) return ak_error_open_file;
    fprintf( fp, "%s\n", message );
#if defined(__unix__) || defined(__APPLE__)
    ak_function_log_syslog( message ); /* все действия дополнительно дублируются в syslog */
#endif
    if( fclose(fp) == EOF ) return ak_error_access_file;
 return ak_error_ok;
}

/* ----------------------------------------------------------------------------------------------- */
 void akrypt_set_audit( TCHAR *message )
{
  if( ak_ptr_is_equal( "stderr", message, 6 )) {
       audit = ak_function_log_stderr; /* если задан stderr, то используем готовую функцию */
  } else {
            if( strlen( message ) > 0 ) {
              memset( audit_filename, 0, 1024 );
              strncpy( audit_filename, message, 1022 );
              audit = akrypt_audit_function;
            }
         }
}

/* ----------------------------------------------------------------------------------------------- */
 int akrypt_file_or_directory( const TCHAR *filename )
{
 struct stat st;

  if( stat( filename, &st )) return 0;
  if( S_ISREG( st.st_mode )) return DT_REG;
  if( S_ISDIR( st.st_mode )) return DT_DIR;

 return 0;
}

/* ----------------------------------------------------------------------------------------------- */
/* выполнение однотипной процедуры с группой файлов */
 int akrypt_find( const char *root , const char *mask,
                                          ak_function_find *function, ak_pointer ptr, bool_t tree )
{
  int error = ak_error_ok;

#ifdef _WIN32
  WIN32_FIND_DATA ffd;
  TCHAR szDir[MAX_PATH];
  char filename[MAX_PATH];
  HANDLE hFind = INVALID_HANDLE_VALUE;
  size_t rlen = 0, mlen = 0;

 #ifdef _MSC_VER
  if( FAILED( StringCchLength( root, MAX_PATH-1, &rlen )))
    return ak_error_message_fmt( ak_error_wrong_length, __func__ ,
                                                             "incorrect length for root variable" );
  if( FAILED( StringCchLength( mask, MAX_PATH-1, &mlen )))
    return ak_error_message_fmt( ak_error_wrong_length, __func__ ,
                                                            "incorrect length for mask variable" );
 #else
  rlen = strlen( root ); mlen = strlen( mask );
 #endif

  if( rlen > (MAX_PATH - ( mlen + 2 )))
    return ak_error_message_fmt( ak_error_wrong_length, __func__ , "directory path too long" );

 #ifdef _MSC_VER
  if( FAILED( StringCchCopy( szDir, MAX_PATH-1, root )))
    return ak_error_message_fmt( ak_error_wrong_length, __func__ ,
                                                            "incorrect copying of root variable" );
  if( FAILED( StringCchCat( szDir, MAX_PATH-1, TEXT( "\\" ))))
    return ak_error_message_fmt( ak_error_wrong_length, __func__ ,
                                                      "incorrect copying of directory separator" );
  if( FAILED( StringCchCat( szDir, MAX_PATH-1, mask )))
    return ak_error_message_fmt( ak_error_wrong_length, __func__ ,
                                                            "incorrect copying of mask variable" );
 #else
  ak_snprintf( szDir, MAX_PATH-1, "%s\\%s", root, mask );
 #endif

 /* начинаем поиск */
  if(( hFind = FindFirstFile( szDir, &ffd )) == INVALID_HANDLE_VALUE )
    return ak_error_message_fmt( ak_error_access_file, __func__ , "given mask search error" );

  do {
       if( ffd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY ) {

         if( !strcmp( ffd.cFileName, "." )) continue;  // пропускаем себя и каталог верхнего уровня
         if( !strcmp( ffd.cFileName, ".." )) continue;

         if( tree ) { // выполняем рекурсию для вложенных каталогов
           memset( szDir, 0, MAX_PATH );
          #ifdef _MSC_VER
           if( FAILED( StringCchCopy( szDir, MAX_PATH-1, root )))
             return ak_error_message_fmt( ak_error_wrong_length, __func__ ,
                                                            "incorrect copying of root variable" );
           if( FAILED( StringCchCat( szDir, MAX_PATH-1, TEXT( "\\" ))))
             return ak_error_message_fmt( ak_error_wrong_length, __func__ ,
                                                      "incorrect copying of directory separator" );
           if( FAILED( StringCchCat( szDir, MAX_PATH-1,  ffd.cFileName )))
             return ak_error_message_fmt( ak_error_wrong_length, __func__ ,
                                                                "incorrect copying of file name" );
          #else
           ak_snprintf( szDir, MAX_PATH-1, "%s\\%s", root,  ffd.cFileName );
          #endif

           if(( error = akrypt_find( szDir, mask, function, ptr, tree )) != ak_error_ok )
             ak_error_message_fmt( error,
                                         __func__, "access to \"%s\" directory denied", filename );
         }
       } else {
               if( ffd.dwFileAttributes &FILE_ATTRIBUTE_SYSTEM ) continue;
                 memset( filename, 0, FILENAME_MAX );
                #ifdef _MSC_VER
                 if( FAILED( StringCchCopy( filename, MAX_PATH-1, root )))
                   return ak_error_message_fmt( ak_error_wrong_length, __func__ ,
                                                            "incorrect copying of root variable" );
                 if( FAILED( StringCchCat( filename, MAX_PATH-1, TEXT( "\\" ))))
                   return ak_error_message_fmt( ak_error_wrong_length, __func__ ,
                                                      "incorrect copying of directory separator" );
                 if( FAILED( StringCchCat( filename, MAX_PATH-1,  ffd.cFileName )))
                   return ak_error_message_fmt( ak_error_wrong_length, __func__ ,
                                                               "incorrect copying of file namme" );
                #else
                 ak_snprintf( filename, MAX_PATH-1, "%s\\%s", root,  ffd.cFileName );
                #endif
                 function( filename, ptr );
              }

  } while( FindNextFile( hFind, &ffd ) != 0);
  FindClose(hFind);

// далее используем механизм функций open/readdir + fnmatch
#else
  DIR *dp = NULL;
  struct dirent *ent = NULL;
  char filename[FILENAME_MAX];

 /* открываем каталог */
  errno = 0;
  if(( dp = opendir( root )) == NULL ) {
    if( errno == EACCES ) return ak_error_message_fmt( ak_error_access_file,
                                          __func__ , _("access to \"%s\" directory denied"), root );
    if( errno > -1 ) return ak_error_message_fmt( ak_error_open_file,
                                                                __func__ , "%s", strerror( errno ));
  }

 /* перебираем все файлы и каталоги */
  while(( ent = readdir( dp )) != NULL ) {
    if( ent->d_type == DT_DIR ) {
      if( !strcmp( ent->d_name, "." )) continue;  // пропускаем себя и каталог верхнего уровня
      if( !strcmp( ent->d_name, ".." )) continue;

      if( tree ) { // выполняем рекурсию для вложенных каталогов
        memset( filename, 0, FILENAME_MAX );
        ak_snprintf( filename, FILENAME_MAX, "%s/%s", root, ent->d_name );
        if(( error = akrypt_find( filename, mask, function, ptr, tree )) != ak_error_ok )
          ak_error_message_fmt( error, __func__, _("access to \"%s\" directory denied"), filename );
      }
    } else
       if( ent->d_type == DT_REG ) { // обрабатываем только обычные файлы
          if( !fnmatch( mask, ent->d_name, FNM_PATHNAME )) {
            memset( filename, 0, FILENAME_MAX );
            ak_snprintf( filename, FILENAME_MAX, "%s/%s", root, ent->d_name );
            function( filename, ptr );
          }
       }
  }
  if( closedir( dp )) return ak_error_message_fmt( ak_error_close_file,
                                                                __func__ , "%s", strerror( errno ));
#endif
 return ak_error_ok;
}

/* ----------------------------------------------------------------------------------------------- */
 int ak_file_read_by_lines( const char *filename, ak_file_read_function *function , ak_pointer ptr )
{
  #define buffer_length ( FILENAME_MAX + 160 )

  char ch;
  struct stat st;
  size_t idx = 0, off = 0;
  int fd = 0, error = ak_error_ok;
  char localbuffer[buffer_length];

 /* проверяем наличие файла и прав доступа к нему */
  if(( fd = open( filename, O_RDONLY | O_BINARY )) < 0 )
    return ak_error_message_fmt( ak_error_open_file,
                             __func__, "wrong open file \"%s\" - %s", filename, strerror( errno ));
  if( fstat( fd, &st ) ) {
    close( fd );
    return ak_error_message_fmt( ak_error_access_file, __func__ ,
                              "wrong stat file \"%s\" with error %s", filename, strerror( errno ));
  }

 /* нарезаем входные на строки длиной не более чем buffer_length - 2 символа */
  memset( localbuffer, 0, buffer_length );
  for( idx = 0; idx < (size_t) st.st_size; idx++ ) {
     if( read( fd, &ch, 1 ) != 1 ) {
       close(fd);
       return ak_error_message_fmt( ak_error_read_data, __func__ ,
                                                                "unexpected end of %s", filename );
     }
     if( off > buffer_length - 2 ) {
       close( fd );
       return ak_error_message_fmt( ak_error_read_data, __func__ ,
                          "%s has a line with more than %d symbols", filename, buffer_length - 2 );
     }
    if( ch == '\n' ) {
      #ifdef _WIN32
       if( off ) localbuffer[off-1] = 0;  /* удаляем второй символ перехода на новую строку */
      #endif
      error = function( localbuffer, ptr );
     /* далее мы очищаем строку независимо от ее содержимого */
      off = 0;
      memset( localbuffer, 0, buffer_length );
    } else localbuffer[off++] = ch;
   /* выходим из цикла если процедура проверки нарушена */
    if( error != ak_error_ok ) return error;
  }

  close( fd );
 return error;
}

/* ----------------------------------------------------------------------------------------------- */
/*                                 реализация вывода справки                                       */
/* ----------------------------------------------------------------------------------------------- */
 int akrypt_litehelp( void )
{
  printf(_("akrypt (crypto application based on libakrypt library, version: %s)\n\n"),
                                                                         ak_libakrypt_version( ));
  printf(_("try \"akrypt --help\" to get more information\n"));

 return EXIT_SUCCESS;
}

/* ----------------------------------------------------------------------------------------------- */
 int akrypt_help( void )
{
  printf(_("akrypt (crypto application based on libakrypt library, version: %s)\n"),
                                                                         ak_libakrypt_version( ));
  printf(_("usage \"akrypt command [options] [files]\"\n\n"));
  printf(_("available commands (in short and long forms):\n"));
  printf(_("  i  icode  calculation and/or checking integrity codes\n"));
  printf(_("     show   show useful information\n\n"));
  printf(_("also try:\n"));
  printf(_("  \"akrypt command --help\" to get information about command options\n"));
  printf(_("  \"man akrypt\" to get more information about akrypt programm and some examples\n"));

 return EXIT_SUCCESS;
}

/* ----------------------------------------------------------------------------------------------- */
/*                                                                                       akrypt.c  */
/* ----------------------------------------------------------------------------------------------- */
